be/src/exec/stream/aggregate/agg_state_data.cpp

```diff
@@ -0,0 +1,294 @@
+        }
+    }
+    DCHECK(j == result_chunks.size());
+    return Status::OK();
+}
+
+// Restore the specific group_by_keys + agg_values's detail into the state.
+Status AggStateData::allocate_retract_state(size_t chunk_size, const std::vector<uint8_t>& keys_not_in_map,
+                                            const Column* raw_column, const std::vector<ChunkPtrOr>& result_chunks,
+                                            const Buffer<AggGroupStatePtr>& agg_group_state) const {
+    int32_t j = 0;
+    for (auto i = 0; i < chunk_size; i++) {
+        DCHECK_LT(j, chunk_size);
+        if (keys_not_in_map[i]) {
+            auto& result_chunk_or = result_chunks[j++];
+            if (result_chunk_or.ok()) {
+                auto result_chunk = result_chunk_or.value();
+                auto& result_columns = result_chunk->columns();
+                DCHECK_EQ(result_columns.size(), 1);
+                std::vector<const Column*> columns = {raw_column, result_columns[0].get()};
+                _agg_function->restore_detail(_agg_fn_ctx, i, columns, agg_group_state[i] + _agg_state_offset);
+            } else {
+                auto status = result_chunk_or.status();
+                DCHECK(status.is_end_of_file());
+                if (!status.is_end_of_file()) {
+                    return status;
+                }
+            }
+        }
+    }
+    DCHECK(j == result_chunks.size());
+    return Status::OK();
+}
+
+Status AggStateData::process_chunk(size_t chunk_size, const std::vector<ChunkPtrOr>& result_chunks,
+                                   const Buffer<uint8_t>& keys_not_in_map, const StreamRowOp* ops,
+                                   std::vector<std::vector<const Column*>>& raw_columns,
+                                   const Buffer<AggDataPtr>& agg_group_state) const {
+    auto* columns = (raw_columns[agg_func_id()]).data();
+    for (int i = 0; i < chunk_size; i++) {
+        auto op = ops[i];
+        switch (op) {
+        case StreamRowOp::INSERT:
+        case StreamRowOp::UPDATE_BEFORE:
+            _agg_function->update(_agg_fn_ctx, columns, agg_group_state[i] + _agg_state_offset, i);
+            break;
+        case StreamRowOp::DELETE:
+        case StreamRowOp::UPDATE_AFTER:
+            _agg_function->retract(_agg_fn_ctx, columns, agg_group_state[i] + _agg_state_offset, i);
+            break;
+        }
+    }
+    return Status::OK();
+}
+
+Status AggStateData::finalize_result(size_t chunk_size, const Buffer<AggDataPtr>& agg_group_data, Column* to) const {
+    _agg_function->batch_finalize(_agg_fn_ctx, chunk_size, agg_group_data, _agg_state_offset, to);
+    return Status::OK();
+}
+
+Status AggStateData::finalize_detail_result(size_t chunk_size, const Columns& group_by_columns,
+                                            const Buffer<AggDataPtr>& agg_group_data,
+                                            const StateTable* detail_state_table, Column* to) const {
+    if (is_detail_state_table()) {
```

Comment:
the detail_state_table means operation history?

Reply:
yep, for min/max aggregation, needs detail state to restore last min/max result.


```
 *  - DETAIL_RESULT         : Use detail table data as AggStateData to retract in retract mode and
 *                              use result table data as intermediate result.
 *  - DETAIL_INTERMEDIATE   : Use detail table data as AggStateData to retract in retract mode and
 *                              use intermediate table data as intermediate result.
```